network/src/protocols/hole_punching/component/connection_request.rs

```diff
@@ -0,0 +1,240 @@
+            .await
+        {
+            return StatusCode::ForwardError.with_context(error);
+        }
+
+        let remote_listens = self
+            .message
+            .listen_addrs()
+            .iter()
+            .filter_map(
+                |raw| match Multiaddr::try_from(raw.bytes().raw_data().to_vec()) {
+                    Ok(mut addr) => {
+                        if let Some(peer_id) = extract_peer_id(&addr) {
+                            if peer_id != from_peer_id {
+                                return None;
+                            }
+                        } else {
+                            addr.push(Protocol::P2P(Cow::Borrowed(from_peer_id.as_bytes())));
+                        }
+
+                        match find_type(&addr) {
+                            TransportType::Memory
+                            | TransportType::Onion
+                            | TransportType::Ws
+                            | TransportType::Wss
+                            | TransportType::Tls => None,
+                            TransportType::Tcp => Some(addr),
+                        }
+                    }
+                    Err(_) => None,
+                },
+            )
+            .collect();
+
+        let mut pending_delivered = self.protocol.pending_delivered.write();
+        let now = unix_time_as_millis();
+        pending_delivered.insert(from_peer_id, (remote_listens, now));
+
+        Status::ok()
+    }
+
+    async fn forward_message(&self, self_peer_id: &PeerId, to_peer_id: &PeerId) -> Status {
+        let content = forward_request(self.message, self_peer_id);
+        let new_message = packed::HolePunchingMessage::new_builder()
+            .set(content)
+            .build()
+            .as_bytes();
+        let proto_id = SupportProtocols::HolePunching.protocol_id();
+
+        let target_sid = self
+            .protocol
+            .network_state
+            .peer_registry
+            .read()
+            .get_key_by_peer_id(to_peer_id);
+
+        match target_sid {
+            Some(to_peer) => {
+                debug!(
+                    "target peer {} is found, forward the request to it",
+                    to_peer_id
+                );
+                if let Err(error) = self
+                    .p2p_control
+                    .send_message_to(to_peer, proto_id, new_message)
+                    .await
+                {
+                    StatusCode::ForwardError.with_context(error)
+                } else {
+                    Status::ok()
+                }
+            }
+            None => {
+                debug!(
+                    "target peer {} is not found, broadcast the request to more peers",
+                    to_peer_id
+                );
+                let sid = self.peer;
+                if let Err(error) = self
+                    .p2p_control
+                    .filter_broadcast(
```

Comment:
There is a message storm problem here, and we need to do message de-duplication: must introduce maintain a short-term cache of recently seen message IDs (pair of from/to) and avoid rebroadcasting duplicates.

And instead of full broadcast, use a gossip protocol with limited fanout (e.g., relay to √N peers). This reduces redundant message spread while preserving reachability.

Reply:
Regarding message storms, there is already a rate limit of `session_id + message.item_id`. If it's not enough, the initiator can add a `to` deduplication, and the intermediary can add a `from + to` deduplication.